File: agent-driven-architecture-main/examples/kotlin/spine/src/main/kotlin/adr/spine/ports/Mailbox.kt
// ── spine/ports/mailbox — the barge-in seam (12.1/12.2) ────────────────────
// A file in spine/ports/ with a body is a gate failure (check C11): a port is a
// published contract, not an implementation. The in-memory implementation lives in
// spine/concurrency/in-memory; a real one is SQS, a Kafka topic or a Postgres queue,
// and 8.5 names "the concurrency / ordering policy of the mailbox" as swappable
// behind exactly this contract.
//
// `messages` is a ReceiveChannel and not a `suspend fun take()` ON PURPOSE. A turn
// must be observable WHILE another turn is running, which means a select over
// { a message arrived, the turn settled }. Kotlin's `select` is ATOMIC — a clause
// that loses does not consume — so a message the race discarded is still in the
// channel. A `suspend take()` would have to be wrapped in a Deferred to be
// selectable, and a Deferred that loses HAS consumed. That difference is the whole
// reason 12.3's drain loop could not preempt.
//
// take LEASES; the message is not gone until `ack`. `ack` is called only AFTER the
// commit (12.2), so a crash between the two re-delivers rather than loses — and the
// durable policy's key dedupe is what makes that redelivery safe.

package adr.spine.ports

import adr.spine.pure.Message
import kotlinx.coroutines.channels.ReceiveChannel

interface Mailbox {
    /** Offer a message. Never blocks the poster on a turn. */
    fun post(message: Message)

    /** The lease stream. Receiving a message does NOT remove it — only [ack] does. */
    val messages: ReceiveChannel<Message>

    /** Called ONLY after the step that consumed the message has committed (12.2). */
    fun ack(message: Message)
}

File: agent-driven-architecture-main/examples/typescript/src/spine/ports/mailbox.ts
// ── spine/ports/mailbox — the barge-in seam (12.2) ─────────────────────────
// INTERFACES ONLY (C11).
//
// `take()` LEASES. The message is not removed from the queue until `ack`, and
// `ack` is called ONLY AFTER THE COMMIT — so a crash between the take and the
// commit re-delivers rather than loses. That is the whole of 12.2's durable
// contract, and it is a property of this signature rather than of a convention
// someone remembers.
//
// `take()` returns a promise so that a consumer can RACE it against the turn it
// is already running. That race is the fix for the drain loop that cannot
// preempt: a blocking take at the top of a loop body is never reached while a
// turn is in flight, so every mid-turn guard downstream of it is dead code.

import type { Message } from "../pure/mailbox";

export interface Mailbox {
  post(message: Message): void;
  /** leases the next message; it stays in flight until `ack` */
  take(): Promise<Message>;
  /** called only after the commit — never before */
  ack(message: Message): void;
}
